File: Reqnroll/Chapter17/17.2-InPlacePersonaPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(
    OrderingContext orderingContext,
    OrderingApiDriver orderingApiDriver,
    BackdoorApiDriver backdoorApiDriver)
{
    [Given("the customer has the following orders")]
    public async Task GivenTheCustomerHasTheFollowingOrders(DataTable ordersTable)
    {
        var orders = ordersTable.CreateSet<NamedOrderData>().ToList();
        foreach (var order in orders)
        {
            var orderRequest = new PlaceOrderRequestObjectMother().Build();
            var placedOrder = await backdoorApiDriver
                .PrepareOrder(DomainDefaults.CustomerName, orderRequest, order.Status)
                .Execute();
            orderingContext.NamedOrders[order.OrderName] = placedOrder.OrderNo;
        }
    }

    [When("they cancel {order}")]
    public async Task WhenTheyCancelOrder(int orderNo)
    {
        await orderingApiDriver.CancelOrder(orderNo).Execute();
    }

    [Then("their order list should contain")]
    public async Task ThenTheirOrderListShouldContain(DataTable expectedOrdersTable)
    {
        var expectedOrders = expectedOrdersTable.CreateSet<NamedOrderData>().ToList();
        foreach (var expectedOrder in expectedOrders)
        {
            if (!orderingContext.NamedOrders.TryGetValue(expectedOrder.OrderName, out int orderNo))
            {
                throw new InvalidOperationException($"Order {expectedOrder.OrderName} not known");
            }

            var order = await orderingApiDriver.GetOrder(orderNo).Execute();
            Assert.AreEqual(expectedOrder.Status, order.Status,
                $"Unexpected status for order {expectedOrder.OrderName}.");
        }
    }
}

File: CucumberJVM/Chapter17/17.2-InPlacePersonaPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.specs.drivers.BackdoorApiDriver;
import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.NamedOrderData;
import com.wimp.app.specs.support.OrderingContext;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final OrderingApiDriver orderingApiDriver;
    private final BackdoorApiDriver backdoorApiDriver;

    public OrderingStepDefinitions(OrderingContext orderingContext, OrderingApiDriver orderingApiDriver, BackdoorApiDriver backdoorApiDriver) {
        this.orderingContext = orderingContext;
        this.orderingApiDriver = orderingApiDriver;
        this.backdoorApiDriver = backdoorApiDriver;
    }

    @Given("the customer has the following orders")
    public void theCustomerHasTheFollowingOrders(List<NamedOrderData> orders) throws Exception {
        for (final NamedOrderData order : orders) {
            var orderRequest = new PlaceOrderRequestObjectMother().build();
            var placedOrder = backdoorApiDriver
                .prepareOrder(DomainDefaults.CUSTOMER_NAME, orderRequest, order.getStatus())
                .execute();
            orderingContext.getNamedOrders().put(order.getOrderName(), placedOrder.getOrderNo());
        }
    }

    @When("they cancel {order}")
    public void theyCancelOrder(int orderNo) throws Exception {
        orderingApiDriver.cancelOrder(orderNo).execute();
    }

    @Then("their order list should contain")
    public void thenTheirOrderListShouldContain(List<NamedOrderData> expectedOrders) throws Exception {
        for (final NamedOrderData expectedOrder : expectedOrders) {
            var orderNo = orderingContext.getNamedOrders().getOrDefault(expectedOrder.getOrderName(), null);
            if (orderNo == null) {
                throw new RuntimeException("Order %s not known".formatted(expectedOrder.getOrderName()));
            }

            var order = orderingApiDriver.getOrder(orderNo).execute();
            assertEquals(expectedOrder.getStatus(), order.getStatus(), "Unexpected status for order %s.".formatted(expectedOrder.getOrderName()));
        }
    }
}
